python/tank_vendor/shotgun_authentication/interactive_authentication.py

```diff
@@ -0,0 +1,432 @@
+        engine, it will run the code in the main thread.
+        :param hostname: Host to display in the dialog.
+        :param login: login to display in the dialog.
+        :param http_proxy: Proxy server to use when validating credentials. Can be None.
+        :returns: A tuple of (hostname, login, session_token)
+        """
+        from .ui import login_dialog
+
+        if self._is_session_renewal:
+            logger.debug("Requesting password in a dialog.")
+        else:
+            logger.debug("Requesting username and password in a dialog.")
+
+        def _process_ui():
+            dlg = login_dialog.LoginDialog(
+                "Shotgun Login",
+                is_session_renewal=self._is_session_renewal,
+                hostname=hostname,
+                login=login,
+                http_proxy=http_proxy,
+                fixed_host=self._fixed_host
+            )
+            return dlg.result()
+
+        result = self._gui_launcher(_process_ui)
+
+        if not result:
+            self.raise_authentication_cancelled_error()
+        return result
+
+
+_renew_session_lock = threading.Lock()
+"""
+Lock the assures only one thread at a time can execute the authentication logic.
+"""
+_renew_session_disabled = False
+"""
+Flag that keeps track if a user cancelled authentication. When the flag is raised, it will
+be impossible to authenticate again.
+"""
+
+
+def _renew_session(user, session_token, credentials_handler):
+    """
+    Common login logic, regardless of how we are actually logging in. It will first try to reuse
+    any existing session and if that fails then it will ask for credentials and upon success
+    the credentials will be cached.
+    :raises: TankAuthenticationError Thrown if the authentication is cancelled.
+    :raises: TankAuthenticationDisabled Thrown if authentication was cancelled before.
+    """
+    logger.debug("About to take the authentication lock.")
+    global _renew_session_lock
+    global _renew_session_disabled
+    with _renew_session_lock:
+
+        # If somebody refreshed the session token on the user object since we tried with
+        # the session token.
+        if user.get_session_token() != session_token:
+            return None
+        logger.debug("Took the authentication lock.")
+        # If somebody disabled authentication, we're done here as well.
+        if _renew_session_disabled:
+            raise AuthenticationDisabled()
+
+        try:
+            logger.debug("Not authenticated, requesting user input.")
+            # Do the actually credentials prompting and authenticating.
+            hostname, login, session_token = credentials_handler.authenticate(
+                user.get_host(),
+                user.get_login(),
+                user.get_http_proxy()
+            )
+        except AuthenticationError:
+            _renew_session_disabled = True
```

Comment:
I cannot seem to find anything that sets this back to `False` once it has been set to true. What's the workflow here if you for press cancel? Rather than being global, perhaps the main authenticator object could be keeping this flag somehow?